.github/workflows/ci.yml

```diff
@@ -34,6 +34,79 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
+      - name: Setting build env
+        run: |
+          echo "CI_OS_NAME=${{ runner.os }}" >> $GITHUB_ENV
+        shell: bash
+
+      # We install gnu-tar because BSD tar is buggy on Github's macos machines.
+      # See <https://github.com/actions/cache/issues/403>.
+      - name: Install GNU tar
+        if: runner.os == 'macOS'
+        run: |
+          brew install gnu-tar
+          echo "/usr/local/opt/gnu-tar/libexec/gnubin" >> $GITHUB_PATH
+
+      # Cache the global cargo directory, but NOT the local `target` directory which
+      # we cannot reuse anyway when the nightly changes (and it grows quite large
+      # over time).
+      - name: Add cache for cargo
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cargo/ci
+            ~/.cargo/registry
+            ~/.cargo/git
+          key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
```

Comment:
So, this means we re-build xargo and rustup-toolchain-install-master each time a lock file changes? That seems strange to me.

Reply:
yeah, that's unfortunate. actions/cache caches artifacts base on this key only.
I haven't found a way to refresh xargo cache
if xargo is updated.

Reply:
I guess the alternative would be to just cache based on the OS -- that's what we did before.

But our lock file does not change so often, so this is fine as well.

Reply:
I'm not sure on appveyor, but travis calculates the checksums by hashing dir contents to find if the cache artifacts changed.
actions/cache of gh actions doesn't do that.

Reply:
Travis reused the cache all the time I am pretty sure, now exactly what they would be hashing in.

Reply:
yes, travis do. They restores the cache each builds base on cache path.
After all the build steps finished, they calculate checksum of dir content to find differences
and refresh cache contents.

Reply:
> They restores the cache each builds base on cache path.

I assume GHA does the same, what else would caching even mean?

> After all the build steps finished, they calculate checksum of dir content to find differences
and refresh cache contents.

Not sure what that has to do with anything, that's just a detail of how the cache gets updated at the end of the build.

Reply:
Or does GHA not automatically update the cache? Does it assume the contents of the cache for a given key never change? I tried to figure this out from the docs but they do not seem to say. That would be a really user-unfriendly caching model I think...

Reply:
actions/cache doesn't do the second step link travis.
You could read their documentation if you want:
- https://docs.github.com/en/free-pro-team@latest/actions/guides/caching-dependencies-to-speed-up-workflows#about-caching-workflow-dependencies
- https://docs.travis-ci.com/user/caching/

Reply:
Reading through [these docs](https://docs.github.com/en/free-pro-team@latest/actions/guides/caching-dependencies-to-speed-up-workflows), it seems indeed like GHA caches are never updated, only created initially. Wow, that changes everything...

Why is GHA so much more complicated to use than anything else?^^

Reply:
Okay so we cannot cache `.rustup` and we'll redownlaod stable from the rust-lang servers each time. This design choice of GHA will lead to hugely increased traffic costs for rust-lang (and other installers), which is horrible -- IMO it would be their responsibility to minimize the impact of CI on other people's infrastructure, and they seem to fail to meet that responsibility.

But there's not much we can do there, so whatever.

Reply:
If you want, I could use https://github.com/actions-rs/toolchain/blob/master/README.md#outputs
and use rustc_hash as a key to refresh .rustup cache.

Reply:
We're not using actions-rs things here at all, are we? Maybe we should? Clippy seems to do that.

We are instead getting our rustup from who-knows-where.^^